### PR TITLE
jobs: fix race which permitted concurrent execution

### LINKED ISSUE #68315: jobs: TestRegistryLifecycle failed
jobs.TestRegistryLifecycle [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3250881&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3250881&tab=artifacts#/) on master @ [c995342ead51e08f8ed1155de4218d30a00d86d2](https://github.com/cockroachdb/cockroach/commits/c995342ead51e08f8ed1155de4218d30a00d86d2):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x516ba00]

```
Stack: 

```
goroutine 74085 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0021d6120, 0x8ea0ad0, 0xc0042069c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x15d
panic(0x5b5e540, 0xad5a700)
	/usr/local/go/src/runtime/panic.go:971 +0x499
github.com/cockroachdb/cockroach/pkg/util/tracing/collector.(*Iterator).Valid(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/collector/collector.go:107
github.com/cockroachdb/cockroach/pkg/sql.glob..func75(0x8ea0ad0, 0xc004206540, 0x8fd05e8, 0xc0042f4568, 0xc00109e598, 0x0, 0x0, 0xc0019bab00, 0xc0045d0960, 0xc0020d6aa0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:1228 +0x1e0
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1(0x8e28660, 0xc00320f3e0, 0xc001902d38, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:632 +0x5ae
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3(0x8ea0ad0, 0xc0042069c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120 +0x1aa
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc0021d6120, 0x8ea0ad0, 0xc0042069c0, 0xc0027f4000, 0x0, 0xc002319e30)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0x162
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x28e
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRegistryLifecycle/dump_traces_on_cancel
    jobs_test.go:250: Starting resume
    jobs_test.go:263: Exiting resume
    jobs_test.go:284: Starting FailOrCancel
    jobs_test.go:284: Starting FailOrCancel
    jobs_test.go:299: Exiting OnFailOrCancel
E210802 06:30:21.191981 74082 jobs/adopt.go:327  [n1] 57  job 680969571288383489: adoption completed with error job canceled
E210802 06:30:21.200704 72865 jobs/registry.go:644  [-] 58  error claiming jobs: could not query jobs table: claim-jobs: context canceled
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestRegistryLifecycle PKG=./pkg/jobs TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/sql-schema spaskob
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegistryLifecycle.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### LINKED ISSUE #68593: Multiple resumers can run concurrently in the same job registry
**The problem**

Each node runs a job registry that manages long-running jobs. A registry monitors 
system.jobs table and try to resume some jobs (start their execution) that are in runnable 
states (`running` and `reverting`). The registry keeps an internal state, `adoptedJobs` 
map, that keeps track of the currently executing resumers, where a resumer is a goroutine 
that executes a job. 

When a resumer starts, it registers its job in the internal state to prevent executing 
redundant resumers for the same job. When the job is paused,  the registry unregisters the 
job and cancels the context of the resumer so that the job can be paused. Consequently, 
while the resumer is canceling its execution, unpausing the job can cause the registry to 
start another resumer  concurrently for the same job, which should not happen. The resume 
can start because the internal state `adoptedJobs` was cleared in response to pause request.
In extreme cases, a specific sequence of events can cause multiple resumers to start concurrently.

**Expected behavior**
Only one resumer per job executes at a time in the same registry.

**How to fix it?**
`adoptedJobs` map acts like a lock. Therefore, a job should only be unregistered by the
resumer that registered the job. Currently a job can be unregistered in separate goroutines 
in response to job pause and cancel requests. For these requests, we should only cancel 
the context of the running resumer instead of unregistering the job.

**How can it happen?**
In some specific sequence of events, multiple resumers can begin performing the same job.
One such sequence of events is given below. In this table, time flows from top to bottom, and 
all events belong to the same job. Moreover, the resumers belong to the same job. `Adopt loop`
monitors system.jobs table and start resumers. `Cancel loop` unregisters a job when the job
is paused or canceled. Resumer-3 and Resumer-4 start concurrently, which should not happen.

```
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|    Job    |    Adopt   |  Cancel  |    Can   |  Resumer  |  Resumer  |  Resumer  |  Resumer   |
|   status  |    loop    |   loop   | start a  |     1     |     2     |     3     |      4     |
|  in Table |            |          | resumer? |           |           |           |            |
+===========+============+==========+==========+===========+===========+===========+============+
|  running  |            |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |    Claim   |          |    Yes   |           |           |           |            |
|           |    job     |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |    Mark    |          |          |           |           |           |            |
|           |     job    |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |    No    |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |   Resume   |          |          |           |           |           |            |
|           |     job    |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |          | Start job |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|   pause-  |            |          |          |           |           |           |            |
| requested |            |          |          |           |           |           |            |
|  (due to  |            |          |          |           |           |           |            |
|  a pause  |            |          |          |           |           |           |            |
|  request) |            |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |   Pick   |          |           |           |           |            |
|           |            |    job   |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |  Unmark  |          |           |           |           |            |
|           |            | job and  |          |           |           |           |            |
|           |            |  start   |          |           |           |           |            |
|           |            |  it to   |          |           |           |           |            |
|           |            |  cancel  |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |    Yes   |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|   paused  |            |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |          |   Start   |           |           |            |
|           |            |          |          | canceling |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|  running  |            |          |          |           |           |           |            |
|  (due to  |            |          |          |           |           |           |            |
|  unpause) |            |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |  Claim job |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |  Mark job  |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |    No    |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           | Resume job |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |          |           | Start job |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |          |   Unmark  |           |           |            |
|           |            |          |          |    job    |           |           |            |
|           |            |          |          |    and    |           |           |            |
|           |            |          |          |   cancel  |           |           |            |
|           |            |          |          |   others  |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |          |           |   Start   |           |            |
|           |            |          |          |           | canceling |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |    Yes   |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |          |  Finished |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |  Claim job |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |  Mark job  |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |    No    |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |          |           |   Unmark  |           |            |
|           |            |          |          |           |  job and  |           |            |
|           |            |          |          |           |  cancel   |           |            |
|           |            |          |          |           |   others  |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |    Yes   |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           | Resume job |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |          |           |           |*Start job*|            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |  Claim job |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |  Mark job  |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |    No    |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           | Resume job |          |          |           |           |           |            |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
|           |            |          |          |           |           |           |*Start job* |
+-----------+------------+----------+----------+-----------+-----------+-----------+------------+
```

### PR DESCRIPTION
**Pull request by adityamaru**
The race was that we'd unregister a job from the `adoptedJobs` map while
processing the pause or cancel requests. This is problematic because the
job also unregisters itself in such cases. However, because the job was
already unregistered, the entry which the now canceled job removes may in
fact correspond to a new execution. We remove this hazard by centralizing
the location where the entry is actually removed. Instead, we have the
cancelation/pause processing only cancel the context of the relevant
execution.

Fixes #68593

Release note: None

Release justification: bug fixes and low-risk updates to new functionality

Previously, TraceCollector.StartIter would return a nil object
if we failed to resolve nodeliveness during initialization. This
led to a NPE.

This change now return a TraceCollector instance with the error field
set to the appropriate error, so that the validity check on the iterator
can correctly handle this scenario.

This change also reworks the dump trace on job cancellation test.
Job cancellation semantics under stress are slightly indeterministic in
terms of how many times the execution of OnFailOrCancel is resumed. This makes
it hard to coordinate when to check and how many trace files to expect.

Fixes: #68315

Release note: None

Release justification: low risk, high benefit changes to existing functionality

### CODE REVIEW

File: pkg/jobs/jobs_test.go [hunk 1]
```diff
@@ -1093,42 +1099,47 @@ func TestRegistryLifecycle(t *testing.T) {
 	})
 
 	t.Run("dump traces on cancel", func(t *testing.T) {
+		skip.UnderStress(t, "job cancellation semantics is flaky under stress")
```
Comment (pbardea) [inline pkg/jobs/jobs_test.go:1102]: If it's flaky due to retries, does it make sense to check that at least a certain number of trace files are there?

File: pkg/util/tracing/collector/collector.go [hunk 2]
```diff
@@ -91,7 +91,7 @@ func (t *TraceCollector) StartIter(ctx context.Context, traceID uint64) *Iterato
 	tc := &Iterator{ctx: ctx, traceID: traceID, collector: t}
 	tc.liveNodes, tc.iterErr = nodesFromNodeLiveness(ctx, t.nodeliveness)
 	if tc.iterErr != nil {
```
Comment (pbardea) [inline pkg/util/tracing/collector/collector.go:93]: Should this method be swallowing this error at all? Is it worthwhile to surface?

**Review by pbardea — COMMENTED** [review]
Do you know why we're failing to resolve node liveness? From a brief skim, it looks like we'd only fail if the keys weren't there or are otherwise invalid -- is this expected in tests?

**Comment by adityamaru** [comment]
>Do you know why we're failing to resolve node liveness? From a brief skim, it looks like we'd only fail if the keys weren't there or are otherwise invalid -- is this expected in tests?

I think it was failing in situations where the call to `maybeDumpTrace` was being invoked after the test had returned. This is actually the same shortcoming that is preventing me from getting the test with cancel to successfully pass under stress. We want to wait for all "resumers" to complete the execution of `maybeDumpTrace` but we don't know how many such resumers are going to be there to be able to wait for all of them.  Checking for `StatusCanceled` at the end does not give us the guarantee that all job resumers have returned from the method. I tried a few approaches one of them is counting how many resumers update the job progress with a new trace ID so that we can wait for those many instances to return before we check the trace files, but there is another issue.

Since some of the resumers get cancelled during `OnFailOrCancel` the order in which https://github.com/adityamaru/cockroach/blob/8a253659275d9e107e073eec4401f5146c647a28/pkg/jobs/jobs_test.go#L298 is set is indeterministic, so we don't know whether to expect a true/false value.

Concretely,
resumer1 <- start OnFailOrCancel
resumer2 <- start OnFailOrCancel
resumer2 <- finish OnFailOrCancel (sets the field to true)
resumer2 <- finish OnFailOrCancel with ctx cancelled (sets the field to false)

**Review by ajwerner — COMMENTED** [review]
> Reviewed 1 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68374)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @dt)

___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 3169 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68374#-Mi6qjOq9xOEzel7Khd9:-Mi6qjOq9xOEzel7KhdA:b-mbaxsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e949fa60d8a91eaff0a79a031841362366e2c72/pkg/ccl/changefeedccl/changefeed_test.go#L3169)):*
> ```
> 				unblock()
> 				waitForNoRecord()
> ```
what is this about? Maybe the deal is that we remove the `unblock` channel from `waitForBlocked` and then make it listen for context cancelation.

**Review by ajwerner — COMMENTED** [review]
Thanks for letting me ride on this PR as opposed to making a separate one!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68374#-:-Mi6rbli1kI1RgUdNXej:bsw6nyi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @dt)

**Review by pbardea — APPROVED** [review]

**Review by ajwerner — APPROVED** [review]
:lgtm_strong:

> Reviewed 1 of 10 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68374#-:-Mi7jzPtADAcpLM5YP27:b-jbsf8z)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @dt, and @pbardea)

**Comment by adityamaru** [comment]
bors r=pbardea,ajwerner

**Pull request merged**